File: android/src/main/java/com/flutterzoom/videosdk/FlutterZoomVideoSdkAudioHelper.java
package com.flutterzoom.videosdk;

import android.app.Activity;
import androidx.annotation.NonNull;

import com.flutterzoom.videosdk.convert.*;

import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import us.zoom.sdk.ZoomVideoSDK;
import us.zoom.sdk.ZoomVideoSDKAudioHelper;
import us.zoom.sdk.ZoomVideoSDKUser;

public class FlutterZoomVideoSdkAudioHelper {

    private Activity activity;

    FlutterZoomVideoSdkAudioHelper(Activity activity) {
        this.activity = activity;
    }

    private ZoomVideoSDKAudioHelper getAudioHelper() {
        ZoomVideoSDKAudioHelper audioHelper = null;
        try {
            audioHelper = ZoomVideoSDK.getInstance().getAudioHelper();
            if (audioHelper == null) {
                throw new Exception("No Audio Helper Found");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return audioHelper;
    }

    public void canSwitchSpeaker(@NonNull MethodChannel.Result result) {
        result.success(getAudioHelper().canSwitchSpeaker());
    }

    public void getSpeakerStatus(@NonNull MethodChannel.Result result) {
        result.success(getAudioHelper().getSpeakerStatus());
    }

    public void muteAudio(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> args = call.arguments();
        String userId = (String) args.get("userId");
        ZoomVideoSDKUser user = FlutterZoomVideoSdkUser.getUser(userId);
        if (user != null) {
            activity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().muteAudio(user)));
                }
            });
        }
    }

    public void unmuteAudio(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> args = call.arguments();
        String userId = (String) args.get("userId");
        ZoomVideoSDKUser user = FlutterZoomVideoSdkUser.getUser(userId);
        if (user != null) {
            activity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().unMuteAudio(user)));
                }
            });
        }
    }

    public void setSpeaker(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> args = call.arguments();
        boolean isOn = (Boolean) args.get("isOn");

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().setSpeaker(isOn)));
            }
        });
    }

    public void startAudio(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().startAudio()));
            }
        });
    }

    public void stopAudio(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().stopAudio()));
            }
        });
    }

    public void subscribe(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().subscribe()));
            }
        });
    }

    public void unsubscribe(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().unSubscribe()));
            }
        });
    }

    public void resetAudioSession(@NonNull MethodChannel.Result result) {
        result.success(false);
    }

    public void cleanAudioSession(@NonNull MethodChannel.Result result) {
        result.success(null);
    }

}

File: ios/Classes/FlutterZoomVideoSdkAudioHelper.m
#import "FlutterZoomVideoSdkAudioHelper.h"
#import "FlutterZoomVideoSdkUser.h"
#import "JSONConvert.h"
#import <AVFAudio/AVAudioSession.h>

@implementation FlutterZoomVideoSdkAudioHelper

- (ZoomVideoSDKAudioHelper *)getAudioHelper {
    ZoomVideoSDKAudioHelper* audioHelper = nil;
    @try {
        audioHelper = [[ZoomVideoSDK shareInstance] getAudioHelper];
        if (audioHelper == nil) {
            NSException *e = [NSException exceptionWithName:@"NoAudioHelperFound" reason:@"No Audio Helper Found" userInfo:nil];
            @throw e;
        }
    } @catch (NSException *e) {
        NSLog(@"%@ - %@", e.name, e.reason);
    }
    return audioHelper;
}

-(void) canSwitchSpeaker: (FlutterResult) result {
    // There's no API for retrieving available outputs other than external devices
    // since iOS devices always have "Speaker".
    result(@YES);
}

-(void) getSpeakerStatus: (FlutterResult) result {
    for (AVAudioSessionPortDescription *output in [[[AVAudioSession sharedInstance] currentRoute] outputs]) {
        if ([[output portType] isEqualToString:@"Speaker"]) {
            result(@YES);
        } else {
            result(@NO);
        }
    }
}

-(void) setSpeaker:(FlutterMethodCall *)call withResult:(FlutterResult) result {
    NSError *error;
    AVAudioSessionPortOverride override =  [call.arguments[@"isOn"] boolValue] ? AVAudioSessionPortOverrideSpeaker : AVAudioSessionPortOverrideNone;
    [[AVAudioSession sharedInstance] overrideOutputAudioPort:override error:&error];
    [[AVAudioSession sharedInstance] setActive:YES error:nil];
    if (error) {
        NSLog(@"%@", error);
    } else {
        result(@"ZoomVideoSDKError_Success");
    }
}

-(void) startAudio: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] startAudio])]);
    });
}

-(void) stopAudio: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] stopAudio])]);
    });
}

-(void) muteAudio:(FlutterMethodCall *)call withResult:(FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        ZoomVideoSDKUser *user = [FlutterZoomVideoSdkUser getUser: call.arguments[@"userId"]];
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] muteAudio:user])]);
    });
}

-(void) unMuteAudio:(FlutterMethodCall *)call withResult:(FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        ZoomVideoSDKUser *user = [FlutterZoomVideoSdkUser getUser: call.arguments[@"userId"]];
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] unmuteAudio:user])]);
    });
}

-(void) subscribe: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] subscribe])]);
    });
}

-(void) unSubscribe: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] unSubscribe])]);
    });
}

-(void) resetAudioSession: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        if ([[self getAudioHelper] resetAudioSession]) {
            result(@YES);
        } else {
            result(@NO);
        }
    }
  );
}

-(void) cleanAudioSession: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        [[self getAudioHelper] cleanAudioSession];
    });
}

@end
